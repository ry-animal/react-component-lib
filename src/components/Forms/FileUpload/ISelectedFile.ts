export interface ISelectedFile {
  // TODO remove arrayBuffer once package updated in web store
  arrayBuffer: ArrayBuffer;
  file: File;
  mimeType: string;
  name: string;
}
