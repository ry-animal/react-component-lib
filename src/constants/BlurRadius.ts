export const BlurRadius = { b15: "15px", b8: "8px" };
