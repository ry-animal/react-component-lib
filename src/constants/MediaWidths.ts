export const MEDIA_WIDTHS = {
  upToExtraSmall: "500px",
  upToSmall: "768px",
  upToMedium: "1024px",
  upToLarge: "1280px",
  upToExtraLarge: "1600px",
};
