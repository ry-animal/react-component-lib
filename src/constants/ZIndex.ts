export const ZIndex = {
  ACCOUNT_PREVIEW: 100,
  SEARCH_RESULTS: 100,
  MOBILE_MENU: 100,
};
