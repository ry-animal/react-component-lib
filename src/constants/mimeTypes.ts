export enum MIME_TYPES {
  APPLICATION_ZIP = "application/zip",
  IMAGE_GIF = "image/gif",
  IMAGE_JPEG = "image/jpeg",
  IMAGE_PNG = "image/png",
  IMAGE_WEBP = "image/webp",
  TEXT_PLAIN = "text/plain",
  TEXT_HTML = "text/html",
  VIDEO_MP4 = "video/mp4",
  VIDEO_WEBM = "video/webm",
  MODEL_GTLF_DASH_BINARY = "model/gtlf-binary",
  MODEL_GTFL_PLUS_JSON = "model/gtfl+json",
}
