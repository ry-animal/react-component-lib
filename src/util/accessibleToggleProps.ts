export function accessibleToggleProps(
  label: string,
  controls: string,
  expanded: boolean,
  popupType?: "menu"
) {
  return {
    "aria-controls": controls,
    "aria-expanded": expanded,
    "aria-haspopup": popupType || true,
    "aria-label": label,
    "aria-pressed": expanded,
    role: "button",
  };
}
