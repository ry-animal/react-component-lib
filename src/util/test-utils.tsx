import { render, RenderOptions } from "@testing-library/react";
import "jest-styled-components";
import * as React from "react";
import { BrowserRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import { LIGHT_THEME } from "../theme";

interface ProvidersProps {
  children: React.ReactNode;
}

function Providers({ children }: ProvidersProps) {
  return (
    <BrowserRouter>
      <ThemeProvider theme={LIGHT_THEME}>{children}</ThemeProvider>
    </BrowserRouter>
  );
}

const customRender = (
  ui: JSX.Element,
  options?:
    | RenderOptions<
        typeof import("@testing-library/dom/types/queries"),
        HTMLElement
      >
    | undefined
) => render(ui, { wrapper: Providers, ...options });

// Duplicated instead of import "../../__mocks__/matchMediaMock"; because jest has trouble resolving imports
Object.defineProperty(window, "matchMedia", {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(),
    removeListener: jest.fn(),
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

// Re-export everything
export * from "@testing-library/react";
// Override render method
export { customRender as render };
