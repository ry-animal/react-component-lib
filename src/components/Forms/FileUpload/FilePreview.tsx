import { memo } from "react";
import { DefaultFile } from ".";
import FolderIcon from "../../../assets/icons/icon-folder.svg";
import { makeDataUrl } from "../../../util/image";
import { ModelViewer } from "../../ModelViewer";
import * as Styled from "./FilePreview.styled";
import STRINGS from "./index.strings";
import { ISelectedFile } from "./ISelectedFile";

const determineMode = (mimeType?: string) => {
  let mode: "image" | "model" | "video" | "unknown" = "unknown";
  if (mimeType?.startsWith("image/")) {
    mode = "image";
  } else if (mimeType?.startsWith("video/")) {
    mode = "video";
  } else if (mimeType?.startsWith("model/")) {
    mode = "model";
  }
  return mode;
};

type Props = {
  defaultValue?: DefaultFile;
  file?: ISelectedFile;
  blur?: boolean;
};

function FilePreviewComponent({ defaultValue, file, blur }: Props) {
  const mimeType = file?.mimeType || defaultValue?.mimeType;
  const mode = determineMode(mimeType);
  const media = file ? makeDataUrl(file) : defaultValue?.mediaUri ?? "";
  return (
    <>
      <Styled.BlurWrapper blur={!!blur} data-testid="blur">
        {mode === "image" && (
          <Styled.ImagePreview alt={STRINGS.preview} src={media} />
        )}
        {mode === "model" && (
          <ModelViewer
            src={media}
            style={{
              maxHeight: Styled.MAX_DIMENSION_PX,
              maxWidth: Styled.MAX_DIMENSION_PX,
            }}
          />
        )}
        {mode === "video" && (
          <Styled.VideoPreview autoPlay controls loop>
            <source src={media} type={mimeType} />
          </Styled.VideoPreview>
        )}
        {mode === "unknown" && (
          <Styled.FileWithoutPreview
            alt={STRINGS.fileSelected}
            src={FolderIcon}
          />
        )}
      </Styled.BlurWrapper>
      <Styled.Name>{file?.name || defaultValue?.name}</Styled.Name>
    </>
  );
}

export const FilePreview = memo(FilePreviewComponent);
