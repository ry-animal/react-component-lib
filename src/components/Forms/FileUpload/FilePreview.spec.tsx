import { BlurRadius } from "../../../constants/BlurRadius";
import { render, screen } from "../../../util/test-utils";
import { FilePreview } from "./FilePreview";
import { ISelectedFile } from "./ISelectedFile";

const sampleImage: ISelectedFile = {
  arrayBuffer: Buffer.from(
    "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mP8/x8AAwMCAO+ip1sAAAAASUVORK5CYII=",
    "base64"
  ),
  file: new File(["hello"], "world"),
  mimeType: "image/png",
  name: "pixel.png",
};

const sampleModel: ISelectedFile = {
  arrayBuffer: Buffer.from(
    "Not a real model! All my samples are too big!",
    "ascii"
  ),
  file: new File(["hello"], "world"),
  mimeType: "model/gltf-binary",
  name: "model.glb",
};

const sampleText: ISelectedFile = {
  arrayBuffer: Buffer.from("Hello world!", "ascii"),
  file: new File(["hello"], "world"),
  mimeType: "text/plain",
  name: "message.txt",
};

// Credit: https://github.com/mathiasbynens/small/blob/master/mp4-with-audio.mp4
const sampleVideo: ISelectedFile = {
  arrayBuffer: Buffer.from(
    "AAAAHGZ0eXBpc29tAAACAGlzb21pc28ybXA0MQAAAAhmcmVlAAAC721kYXQhEAUgpBv/wAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA3pwAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAcCEQBSCkG//AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAADengAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAcAAAAsJtb292AAAAbG12aGQAAAAAAAAAAAAAAAAAAAPoAAAALwABAAABAAAAAAAAAAAAAAAAAQAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAADAAAB7HRyYWsAAABcdGtoZAAAAAMAAAAAAAAAAAAAAAIAAAAAAAAALwAAAAAAAAAAAAAAAQEAAAAAAQAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAACRlZHRzAAAAHGVsc3QAAAAAAAAAAQAAAC8AAAAAAAEAAAAAAWRtZGlhAAAAIG1kaGQAAAAAAAAAAAAAAAAAAKxEAAAIAFXEAAAAAAAtaGRscgAAAAAAAAAAc291bgAAAAAAAAAAAAAAAFNvdW5kSGFuZGxlcgAAAAEPbWluZgAAABBzbWhkAAAAAAAAAAAAAAAkZGluZgAAABxkcmVmAAAAAAAAAAEAAAAMdXJsIAAAAAEAAADTc3RibAAAAGdzdHNkAAAAAAAAAAEAAABXbXA0YQAAAAAAAAABAAAAAAAAAAAAAgAQAAAAAKxEAAAAAAAzZXNkcwAAAAADgICAIgACAASAgIAUQBUAAAAAAfQAAAHz+QWAgIACEhAGgICAAQIAAAAYc3R0cwAAAAAAAAABAAAAAgAABAAAAAAcc3RzYwAAAAAAAAABAAAAAQAAAAIAAAABAAAAHHN0c3oAAAAAAAAAAAAAAAIAAAFzAAABdAAAABRzdGNvAAAAAAAAAAEAAAAsAAAAYnVkdGEAAABabWV0YQAAAAAAAAAhaGRscgAAAAAAAAAAbWRpcmFwcGwAAAAAAAAAAAAAAAAtaWxzdAAAACWpdG9vAAAAHWRhdGEAAAABAAAAAExhdmY1Ni40MC4xMDE=",
    "base64"
  ),
  file: new File(["hello"], "world"),
  mimeType: "video/mp4",
  name: "mp4-with-audio.mp4",
};

describe("<FilePreview />", () => {
  global.URL.createObjectURL = jest.fn(
    () => "blob:http://localhost:6006/af2e50bd-bc94-4dff-ae8c-440eabdac5c6"
  );
  test("Default rendering matches snapshot (image file)", () => {
    expect(
      render(<FilePreview file={sampleImage} />).asFragment()
    ).toMatchSnapshot();
  });

  test("Default rendering matches snapshot (model file)", () => {
    expect(
      render(<FilePreview file={sampleModel} />).asFragment()
    ).toMatchSnapshot();
  });

  test("Default rendering matches snapshot (text file)", () => {
    expect(
      render(<FilePreview file={sampleText} />).asFragment()
    ).toMatchSnapshot();
  });

  test("Default rendering matches snapshot (video file)", () => {
    expect(
      render(<FilePreview file={sampleVideo} />).asFragment()
    ).toMatchSnapshot();
  });

  test("Default rendering matches snapshot (no file)", () => {
    expect(render(<FilePreview />).asFragment()).toMatchSnapshot();
  });

  test("Default rendering matches snapshot (blur)", () => {
    expect(
      render(<FilePreview file={sampleImage} blur={true} />).asFragment()
    ).toMatchSnapshot();
  });

  test("Blur Wrapper has blur radius style", async () => {
    render(<FilePreview file={sampleImage} blur={true} />).asFragment();

    const blurWrapper = await screen.findByTestId("blur");

    expect(blurWrapper).toHaveStyleRule("filter", `blur(${BlurRadius.b8})`);
  });
});
