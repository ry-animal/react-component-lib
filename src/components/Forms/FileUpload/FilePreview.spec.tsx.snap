// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilePreview /> Default rendering matches snapshot (blur) 1`] = `
<DocumentFragment>
  .c1 {
  max-height: 200px;
  max-width: 200px;
}

.c0 {
  display: inline-block;
  -webkit-filter: blur(8px);
  filter: blur(8px);
}

<div
    class="c0"
    data-testid="blur"
  >
    <img
      alt="Preview of uploaded image"
      class="c1"
      src="blob:http://localhost:6006/af2e50bd-bc94-4dff-ae8c-440eabdac5c6"
    />
  </div>
  .c0 {
  color: #404040;
}

<div
    class="c0"
  >
    pixel.png
  </div>
</DocumentFragment>
`;

exports[`<FilePreview /> Default rendering matches snapshot (image file) 1`] = `
<DocumentFragment>
  .c1 {
  max-height: 200px;
  max-width: 200px;
}

.c0 {
  display: inline-block;
}

<div
    class="c0"
    data-testid="blur"
  >
    <img
      alt="Preview of uploaded image"
      class="c1"
      src="blob:http://localhost:6006/af2e50bd-bc94-4dff-ae8c-440eabdac5c6"
    />
  </div>
  .c0 {
  color: #404040;
}

<div
    class="c0"
  >
    pixel.png
  </div>
</DocumentFragment>
`;

exports[`<FilePreview /> Default rendering matches snapshot (model file) 1`] = `
<DocumentFragment>
  .c0 {
  display: inline-block;
}

<div
    class="c0"
    data-testid="blur"
  >
    <model-viewer
      alt="3d model"
      autoplay="true"
      camera-controls="false"
      environment-image="[object Object]"
      exposure="2"
      src="blob:http://localhost:6006/af2e50bd-bc94-4dff-ae8c-440eabdac5c6"
      style="display: inline-block; max-height: 200px; max-width: 200px;"
    />
  </div>
  .c0 {
  color: #404040;
}

<div
    class="c0"
  >
    model.glb
  </div>
</DocumentFragment>
`;

exports[`<FilePreview /> Default rendering matches snapshot (no file) 1`] = `
<DocumentFragment>
  .c1 {
  height: 150px;
  width: 150px;
}

.c0 {
  display: inline-block;
}

<div
    class="c0"
    data-testid="blur"
  >
    <img
      alt="A file has been selected"
      class="c1"
      src="[object Object]"
    />
  </div>
  .c0 {
  color: #404040;
}

<div
    class="c0"
  />
</DocumentFragment>
`;

exports[`<FilePreview /> Default rendering matches snapshot (text file) 1`] = `
<DocumentFragment>
  .c1 {
  height: 150px;
  width: 150px;
}

.c0 {
  display: inline-block;
}

<div
    class="c0"
    data-testid="blur"
  >
    <img
      alt="A file has been selected"
      class="c1"
      src="[object Object]"
    />
  </div>
  .c0 {
  color: #404040;
}

<div
    class="c0"
  >
    message.txt
  </div>
</DocumentFragment>
`;

exports[`<FilePreview /> Default rendering matches snapshot (video file) 1`] = `
<DocumentFragment>
  .c1 {
  max-height: 200px;
  max-width: 200px;
}

.c0 {
  display: inline-block;
}

<div
    class="c0"
    data-testid="blur"
  >
    <video
      autoplay=""
      class="c1"
      controls=""
      loop=""
    >
      <source
        src="blob:http://localhost:6006/af2e50bd-bc94-4dff-ae8c-440eabdac5c6"
        type="video/mp4"
      />
    </video>
  </div>
  .c0 {
  color: #404040;
}

<div
    class="c0"
  >
    mp4-with-audio.mp4
  </div>
</DocumentFragment>
`;
